### DISCUSSION TITLE
Restrict @babel/parser to certain node types

### DISCUSSION THREAD

**Discussion by ghost** [babel/babel#13925] [Q&A] [answered]
I've read through the docs a few times now, but I feel like I'm maybe missing the "easy" way to do this, so posting here for a quick sanity check.

Say for example you wanted to parse an expression (using `parseExpression`) but restrict the node types it will accept.  For instance, if I wanted to only accept `BinaryExpressions`, `StringLiterals`, and `NumericLiterals`, is there a simple way to configure that through some type of option or list?

I could create a plugin/`@babel/traverse` and visitors, but wasn't sure if that was overkill or the recommended way to do this.  The ideal output would either be the successfully parsed AST (exactly how it works right now), or throw an error if it encounters a node type that I don't have listed.

**Comment by nicolo-ribaudo** [comment, marked as answer]
There is no way to do this in `@babel/parser`: you need to traverse the full AST.
Rather than using `@babel/traverse`, you could use `traverseFast` from `@babel/types` which is more lightweight.

**Reply by ghost** [reply]
Appreciate the quick reply! Are there docs on that? I'm having trouble finding the actual APIs for those packages; I seem to get stuck in a loop between the following pages:

* https://babeljs.io/docs/en/babel-types
* https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/plugin-handbook.md#babel-types

But can't seem to find actual full set of docs for what the methods are.  Another example of this was trying to find out the options that are accepted for `@babel/traverse` but to no avail :(

Thanks again :)

**Reply by nicolo-ribaudo** [reply]
Yeah, we don't have good docs for plugin authors :sweat_smile: 

The best I can suggest is to check the source: you can find the full list of `@babel/types` exports at https://github.com/babel/babel/blob/main/packages/babel-types/src/index.ts, and the `traverseFast` signature at https://github.com/babel/babel/blob/main/packages/babel-types/src/traverse/traverseFast.ts.

**Reply by nicolo-ribaudo** [reply]
You would need something like this:
```js
const t = require("@babel/types");
const parser = require("@babel/parser");

const ast = parser.parseExpression(code);
t.traverseFast(ast, node => {
  if (!allowedTypes.has(node.type)) throw new Error("Unexpected node");
});
```

**Reply by ghost** [reply]
I definitely understand the struggle, haha :p

Awesome, I was getting hung up on some errors around scope and parent path when I was trying to pass the AST from `parseExpression` to `@babel/traverse` versus using `parse`, etc., but this looks like exactly what I wanted.  Thanks again for the help!
